# Lab: Building an Interactive Dashboard

In the lecture, we built a dashboard one piece at a time. In this lab, you will bring together the analysis workflow, Plotly figures, and Dash components to create an interactive dashboard using a dataset of your choosing.

Your dashboard should help a specific audience answer a focused question. The goal is not to fit every possible chart onto one page. Select a small set of related evidence, organize it clearly, and give the user one useful way to interact with it.



## Learning goals

By the end of this lab, you should be able to:

- use the reproducible analysis workflow to plan a dashboard;
- prepare and check data before displaying it;
- create Plotly figures that answer related questions;
- organize text, controls, and figures in a Dash layout;
- connect an interactive control to figures with a callback; and
- explain the conclusion and limitations of a dashboard.



## Using AI tools responsibly

You may use an AI tool to ask questions, debug code, or receive feedback. You remain responsible for understanding every part of the submitted work, testing the dashboard, citing the dataset, and checking that the evidence supports your conclusion. Do not submit private, confidential, or personally identifying data to an AI system.



## Assignment requirements

Choose a dataset that supports a meaningful comparison. It should contain:

- enough observations to reveal a pattern rather than a few isolated cases;
- at least one categorical feature that can be used as an interactive filter or comparison;
- at least two additional useful features, which may be categorical, quantitative, temporal, spatial, or text-derived; and
- enough documentation for you to explain what the records and selected features represent.

The completed dashboard must contain:

- a descriptive title and a short explanation of its purpose;
- at least two related Plotly figures;
- at least one dropdown menu or radio-button control;
- at least one callback that updates both figures;
- human-readable titles, labels, categories, and colors;
- stable axes or category orders when changing them would make comparisons misleading; and
- a brief conclusion and limitation displayed on the page.



# Step 1: Question



## 1. Identify the audience

A dashboard should be designed for someone. Identify a realistic audience and explain what decision, judgment, or understanding the dashboard could support.



**Who is the intended audience, and why would this evidence be useful to them?**
The intended audience is visitors and tourists traveling to Los Angeles. This evidence would help them compare Airbnb listings based on factors such as price, location, availability, and guest reviews so they can find a quality place to stay in an area that best fits their needs.


## 2. State the research question

Write one focused question that your two figures will help answer. The question should name the main features or groups being compared.



**What question will the dashboard answer?**
How do Airbnb prices and availability vary across Los Angeles neighborhoods and room types?


## 3. Make a prediction



**What pattern do you expect to find, and what informed that prediction?**

I predict that Airbnb listings in popular tourist areas will have a higher average price than listings in areas farther away from major attractions. I expect neighborhoods near destinations such as Hollywood, Santa Monica, and the Americana at Brand to have higher prices because these locations are more desirable for visitors and provide easier access to popular attractions.

# Step 2: Data



## 4. Import the libraries

Import Pandas, Plotly Express, and the Dash objects used in the lecture: `Dash`, `dcc`, `html`, `Input`, and `Output`.



In [1]:
%pip install dash

Note: you may need to restart the kernel to use updated packages.


In [2]:
import pandas as pd
from dash import Dash, dcc, html, Input, Output
import plotly.express as px


## 5. Document the dataset

Give the dataset title, its creator or publisher, a working link or citation, the time period represented, and any relevant collection or sampling information.



**Where did the data come from, and what does its documentation tell you about how it was collected?**

The data comes from [Inside AirBnb](https://insideairbnb.com/get-the-data/), which collects publicly available information from Airbnb listings. According to its documentation, the data is compiled from information displayed on Airbnb's website, including listing details, availability calendars, and reviews. Inside Airbnb then verifies, cleans, analyzes, and aggregates the data for public use. For Los Angeles, the dataset includes summary listing information intended for analysis and visualization.

## 6. Read the data

Read the dataset into a Pandas DataFrame. Give the DataFrame a short, descriptive name.



In [3]:
insidebnb = pd.read_csv('data/AirBnB_listings.csv')
insidebnb.head()


,id,name,host_id,host_profile_id,host_name,neighbourhood_group,neighbourhood,latitude,longitude,room_type,price,minimum_nights,number_of_reviews,last_review,reviews_per_month,calculated_host_listings_count,availability_365,number_of_reviews_ltm,license
0,2708,"Run Runyon, Beaut Furn Mirror Mini-Suit w/ Fir...",3008.0,1.462506e+18,⁨Chas.⁩,City of Los Angeles,Hollywood,34.09625,-118.34605,Private room,67.0,30.0,47,2026-05-26,0.32,2.0,301,3,NaN
1,2732,Zen Life at the Beach,3041.0,1.462506e+18,Yogapriestess,Other Cities,Santa Monica,34.00440,-118.48095,Private room,214.0,1.0,24,2022-08-21,0.13,2.0,350,0,228269
2,6033,Poolside Serenity Studio,11619.0,1.462507e+18,Sarah,City of Los Angeles,Woodland Hills,34.16887,-118.64478,Entire home/apt,100.0,30.0,19,2020-03-22,0.09,7.0,270,0,NaN
3,6931,"RUN Runyon, Beau Furn Rms Terrace Hollyw Hill ...",3008.0,1.462506e+18,⁨Chas.⁩,City of Los Angeles,Hollywood,34.09626,-118.34372,Private room,95.0,30.0,39,2025-12-15,0.19,2.0,259,2,NaN
4,7874,HD Wi-Fi E Bellflower Cerritos Lakewood LB LA,21700.0,1.462507e+18,Henry,Other Cities,Bellflower,33.87687,-118.11444,Private room,113.0,1.0,26,2026-04-02,0.29,4.0,103,9,NaN


## 7. Preview and inspect the DataFrame

Display the first five rows, last five rows, stored data types, and missing-value counts.



In [4]:
insidebnb.head()

,id,name,host_id,host_profile_id,host_name,neighbourhood_group,neighbourhood,latitude,longitude,room_type,price,minimum_nights,number_of_reviews,last_review,reviews_per_month,calculated_host_listings_count,availability_365,number_of_reviews_ltm,license
0,2708,"Run Runyon, Beaut Furn Mirror Mini-Suit w/ Fir...",3008.0,1.462506e+18,⁨Chas.⁩,City of Los Angeles,Hollywood,34.09625,-118.34605,Private room,67.0,30.0,47,2026-05-26,0.32,2.0,301,3,NaN
1,2732,Zen Life at the Beach,3041.0,1.462506e+18,Yogapriestess,Other Cities,Santa Monica,34.00440,-118.48095,Private room,214.0,1.0,24,2022-08-21,0.13,2.0,350,0,228269
2,6033,Poolside Serenity Studio,11619.0,1.462507e+18,Sarah,City of Los Angeles,Woodland Hills,34.16887,-118.64478,Entire home/apt,100.0,30.0,19,2020-03-22,0.09,7.0,270,0,NaN
3,6931,"RUN Runyon, Beau Furn Rms Terrace Hollyw Hill ...",3008.0,1.462506e+18,⁨Chas.⁩,City of Los Angeles,Hollywood,34.09626,-118.34372,Private room,95.0,30.0,39,2025-12-15,0.19,2.0,259,2,NaN
4,7874,HD Wi-Fi E Bellflower Cerritos Lakewood LB LA,21700.0,1.462507e+18,Henry,Other Cities,Bellflower,33.87687,-118.11444,Private room,113.0,1.0,26,2026-04-02,0.29,4.0,103,9,NaN


In [5]:
insidebnb.tail()

,id,name,host_id,host_profile_id,host_name,neighbourhood_group,neighbourhood,latitude,longitude,room_type,price,minimum_nights,number_of_reviews,last_review,reviews_per_month,calculated_host_listings_count,availability_365,number_of_reviews_ltm,license
43927,1707678777570851520,The French Lime Pool House - Sleeps 8 - Air ho...,588675098.0,1.463893e+18,Posh,Unincorporated Areas,Desert View Highlands,34.587350,-118.160880,Entire home/apt,220.0,2.0,0,NaN,NaN,13.0,361,0,NaN
43928,1707701464310658544,Hollywood City Views The Godfrey,383724181.0,1.469787e+18,Upscale LA,Other Cities,West Hollywood,34.093596,-118.343729,Entire home/apt,399.0,3.0,0,NaN,NaN,14.0,361,0,Exempt
43929,1707712658970417218,Spacious Waterfront 3BR Marina Retreat,412269286.0,1.469879e+18,Tee,Unincorporated Areas,Marina del Rey,33.976153,-118.452329,Entire home/apt,927.0,14.0,0,NaN,NaN,1.0,365,0,NaN
43930,1707714248958928308,Zen in Hollywood Walk of Fame,383724181.0,1.469787e+18,Upscale LA,Other Cities,West Hollywood,34.093596,-118.343729,Entire home/apt,532.0,1.0,0,NaN,NaN,14.0,361,0,Exempt
43931,1707753561170835767,West Hills Oasis,382786700.0,1.469784e+18,Azita,Other Cities,Calabasas,34.140982,-118.653587,Entire home/apt,296.0,1.0,0,NaN,NaN,3.0,363,0,NaN


In [7]:
insidebnb.describe()

,id,host_id,host_profile_id,latitude,longitude,price,minimum_nights,number_of_reviews,reviews_per_month,calculated_host_listings_count,availability_365,number_of_reviews_ltm
count,4.393200e+04,4.375100e+04,4.353700e+04,43932.000000,43932.000000,38362.000000,43921.000000,43932.000000,32621.000000,43751.000000,43932.000000,43932.000000
mean,8.626345e+17,1.037017e+16,1.474995e+18,34.054935,-118.311397,383.136124,17.622868,42.253801,1.393066,16.690704,223.725849,8.592666
std,6.427870e+17,1.318502e+17,3.653041e+16,0.148090,0.171437,662.477421,29.788935,91.794272,1.720861,40.128369,126.962549,17.237732
min,2.708000e+03,7.670000e+02,1.462506e+18,33.338540,-118.917134,4.000000,1.000000,0.000000,0.010000,1.000000,0.000000,0.000000
25%,4.729332e+07,3.012027e+07,1.462932e+18,33.997436,-118.412642,116.000000,1.000000,0.000000,0.200000,1.000000,110.000000,0.000000
50%,1.019753e+18,1.410490e+08,1.466939e+18,34.060284,-118.340230,224.000000,7.000000,7.000000,0.750000,3.000000,262.000000,1.000000
75%,1.448199e+18,4.403883e+08,1.469887e+18,34.110380,-118.223161,406.000000,30.000000,41.000000,2.080000,11.000000,342.000000,10.000000
max,1.707754e+18,1.707457e+18,1.707507e+18,34.811180,-117.652908,20583.000000,1124.000000,3240.000000,54.500000,275.000000,365.000000,1027.000000


In [8]:
insidebnb.info()

<class 'pandas.DataFrame'>
RangeIndex: 43932 entries, 0 to 43931
Data columns (total 19 columns):
 #   Column                          Non-Null Count  Dtype  
---  ------                          --------------  -----  
 0   id                              43932 non-null  int64  
 1   name                            43932 non-null  str    
 2   host_id                         43751 non-null  float64
 3   host_profile_id                 43537 non-null  float64
 4   host_name                       43548 non-null  str    
 5   neighbourhood_group             43932 non-null  str    
 6   neighbourhood                   43932 non-null  str    
 7   latitude                        43932 non-null  float64
 8   longitude                       43932 non-null  float64
 9   room_type                       43932 non-null  str    
 10  price                           38362 non-null  float64
 11  minimum_nights                  43921 non-null  float64
 12  number_of_reviews               43932 non-n

In [ ]:
insidebnb.isnull().sum()

id                                    0
name                                  0
host_id                             181
host_profile_id                     395
host_name                           384
neighbourhood_group                   0
neighbourhood                         0
latitude                              0
longitude                             0
room_type                             0
price                              5570
minimum_nights                       11
number_of_reviews                     0
last_review                       11311
reviews_per_month                 11311
calculated_host_listings_count      181
availability_365                      0
number_of_reviews_ltm                 0
license                           30740
dtype: int64

In [10]:
columns_to_drop = [
    "id", # non contributing identifier
    "host_id", # non contributing identifier
    "host_profile_id", # non contributing identifier
    "host_name", # non contributing identifier
    "neighbourhood_group",# might be empty/not useful in this dataset
    "last_review", # has 11,311 missing values
    "license" # has 30,740 missing values. I don't care if the place has a license or not I just wanna sleep comfy!
]

# Data Cleaning!

In [11]:
insidebnb.loc[
    insidebnb['reviews_per_month'].isna(),
    ['number_of_reviews', 'reviews_per_month']
].head(20)

,number_of_reviews,reviews_per_month
46,0,NaN
72,0,NaN
123,0,NaN
211,0,NaN
214,0,NaN
242,0,NaN
282,0,NaN
308,0,NaN
323,0,NaN
376,0,NaN


In [12]:
# cleaned dataframe
insidebnb = insidebnb[
    [
        "name",
        "neighbourhood",
        "latitude",
        "longitude",
        "room_type",
        "price",
        "minimum_nights",
        "number_of_reviews",
        "reviews_per_month",
        "availability_365",
        "number_of_reviews_ltm",
        "calculated_host_listings_count"
    ]
]

In [14]:
insidebnb.dtypes

name                                  str
neighbourhood                         str
latitude                          float64
longitude                         float64
room_type                             str
price                             float64
minimum_nights                    float64
number_of_reviews                   int64
reviews_per_month                 float64
availability_365                    int64
number_of_reviews_ltm               int64
calculated_host_listings_count    float64
dtype: object

**Complete the table for every feature you expect to use. Add or remove rows as needed.**

| Feature | What it represents | Data storage type | Feature type | Missing values | Dashboard role |
|---|---|---|---|---:|---|
| name | name of the bnb listing | string/object | identifier | 0 | hover text / listing identification |
| neighborhood | LA neighborhood where bnb location is listed | string/object | categorical | 0 | main filter and location |
| latitude | geographic coordinate | Float | Quantitative, continuous | 0 | Map/location visualization |
| longitude | geographic coordinate | Float | Quantitative, continuous | 0 | Map/location visualization |
| room_type | Type of accommodation, such as entire home or private room | string | categorical | 0 | filtering/grouping variable |
| price | listing price | float | quantitative | 5570 | main outcome variable |
| minimum_nights | minimum number of nights required for booking | integer | quantitative | 11 | booking-restriction |
| availability_365 | Number of days available to book during the next 365 days  | integer | Quantitative | 0 | availability comparison |
| number_of_reviews_ltm | reviews received in the past year(12 months) | integer | Quantitative | 0 | Availability comparison |
| calculated_host_listings_count | number of listings associated with the host | integer | quantitative | 181 | Host size |




**What does one row of the original dataset represent?**



# Step 3: Operation



## 8. Create the analysis DataFrame

Select the features needed for the dashboard and use `.copy()` to create a DataFrame named `analysis`. Handle missing values in a way that is appropriate for the question. If values need clearer labels or corrected storage types, make those changes here.



In [15]:
analysis = insidebnb[
    [
        "name",
        "neighbourhood",
        "room_type",
        "price",
        "minimum_nights",
        "number_of_reviews",
        "reviews_per_month",
        "availability_365",
        "latitude",
        "longitude"
    ]
].copy()



In [16]:
analysis = analysis.dropna(subset=["price"])

In [ ]:
analysis.loc[
    analysis["reviews_per_month"].isna(),
    "number_of_reviews"
].value_counts()

Series([], Name: count, dtype: int64)

In [32]:
analysis["reviews_per_month"] = analysis["reviews_per_month"].fillna(0)

In [33]:
analysis.isnull().sum()

name                 0
neighbourhood        0
room_type            0
price                0
minimum_nights       9
number_of_reviews    0
reviews_per_month    0
availability_365     0
latitude             0
longitude            0
price_category       0
dtype: int64

**How did you handle missing values, and why was that choice appropriate?**
I removed rows with missing prices because price is one of the main features being analyzed in the dashboard. Listings without a price cannot contribute to price comparisons. Missing reviews_per_month values were replaced with 0 for listings with no reviews because those listings have not received reviews during any month.

## 9. Create any calculated features

If the dashboard needs a calculated category, date component, percentage, rate, or other calculated feature, create it here. If no calculated feature is needed, leave the code cell blank and explain why.



In [19]:
analysis["price_category"] = pd.cut(
    analysis["price"],
    bins=[0, 100, 200, 300, 500, float("inf")],
    labels=[
        "Under $100",
        "$100–199",
        "$200–299",
        "$300–499",
        "$500+"
    ]
)

In [20]:
analysis["price_category"].value_counts()

price_category
$100–199      9862
Under $100    7670
$500+         7090
$300–499      7048
$200–299      6692
Name: count, dtype: int64

**What calculated feature or cleaned label did you create, and how will it support the dashboard?**

I created a price_category feature that groups listings into understandable nightly price ranges. This will help dashboard users quickly compare affordability across Los Angeles neighborhoods.


## 10. Prepare the control options

Create the labels and values for the dropdown or radio-button control. The visible labels should be understandable to someone who has not seen the original column names.



In [ ]:
#checking number of neighborhoods I have
analysis["neighbourhood"].nunique()


264

In [ ]:
analysis["neighbourhood"].value_counts()

neighbourhood
Long Beach        1655
Hollywood         1338
Venice            1268
West Hollywood    1147
Santa Monica      1063
                  ... 
Elizabeth Lake       2
Walnut Park          2
Griffith Park        1
Val Verde            1
Cudahy               1
Name: count, Length: 264, dtype: int64

In [23]:
neighbourhood_options = [
    {"label": neighbourhood, "value": neighbourhood}
    for neighbourhood in sorted(analysis["neighbourhood"].dropna().unique())
]

In [ ]:
neighbourhood_options[:10]

[{'label': 'Acton', 'value': 'Acton'},
 {'label': 'Adams-Normandie', 'value': 'Adams-Normandie'},
 {'label': 'Agoura Hills', 'value': 'Agoura Hills'},
 {'label': 'Agua Dulce', 'value': 'Agua Dulce'},
 {'label': 'Alhambra', 'value': 'Alhambra'},
 {'label': 'Alondra Park', 'value': 'Alondra Park'},
 {'label': 'Altadena', 'value': 'Altadena'},
 {'label': 'Angeles Crest', 'value': 'Angeles Crest'},
 {'label': 'Arcadia', 'value': 'Arcadia'},
 {'label': 'Arleta', 'value': 'Arleta'}]

In [25]:
# checking if price is a valid numeric
analysis["price"].dtype

dtype('float64')

In [26]:
analysis = analysis.dropna(subset=['price'])

In [27]:
analysis.loc[
    analysis["reviews_per_month"].isna(),
    "number_of_reviews"
].value_counts()

Series([], Name: count, dtype: int64)

In [29]:
analysis["neighbourhood"].value_counts().head(20)

neighbourhood
Long Beach              1655
Hollywood               1338
Venice                  1268
West Hollywood          1147
Santa Monica            1063
Downtown                 958
Exposition Park          863
Pasadena                 701
Beverly Hills            669
Glendale                 637
Hollywood Hills West     612
Culver City              607
Alhambra                 607
Hollywood Hills          585
Burbank                  557
Mid-City                 532
Inglewood                464
Woodland Hills           458
Sherman Oaks             448
Beverly Grove            438
Name: count, dtype: int64

In [40]:
# keeping only neighborhoods with at least 300 listings
neighbourhood_counts = analysis["neighbourhood"].value_counts()

valid_neighbourhoods = neighbourhood_counts[
    neighbourhood_counts >= 300
].index

In [31]:
neighbourhood_options = [
    {"label": neighbourhood, "value": neighbourhood}
    for neighbourhood in sorted(valid_neighbourhoods)
]

# Step 4: Check



## 11. Check the prepared data

Display the first rows of `analysis`, its stored data types, missing-value counts, and appropriate summaries or frequency counts for the dashboard features.



In [34]:
analysis.head()


,name,neighbourhood,room_type,price,minimum_nights,number_of_reviews,reviews_per_month,availability_365,latitude,longitude,price_category
0,"Run Runyon, Beaut Furn Mirror Mini-Suit w/ Fir...",Hollywood,Private room,67.0,30.0,47,0.32,301,34.09625,-118.34605,Under $100
1,Zen Life at the Beach,Santa Monica,Private room,214.0,1.0,24,0.13,350,34.00440,-118.48095,$200–299
2,Poolside Serenity Studio,Woodland Hills,Entire home/apt,100.0,30.0,19,0.09,270,34.16887,-118.64478,Under $100
3,"RUN Runyon, Beau Furn Rms Terrace Hollyw Hill ...",Hollywood,Private room,95.0,30.0,39,0.19,259,34.09626,-118.34372,Under $100
4,HD Wi-Fi E Bellflower Cerritos Lakewood LB LA,Bellflower,Private room,113.0,1.0,26,0.29,103,33.87687,-118.11444,$100–199


In [35]:
analysis.info()

<class 'pandas.DataFrame'>
Index: 38362 entries, 0 to 43931
Data columns (total 11 columns):
 #   Column             Non-Null Count  Dtype   
---  ------             --------------  -----   
 0   name               38362 non-null  str     
 1   neighbourhood      38362 non-null  str     
 2   room_type          38362 non-null  str     
 3   price              38362 non-null  float64 
 4   minimum_nights     38353 non-null  float64 
 5   number_of_reviews  38362 non-null  int64   
 6   reviews_per_month  38362 non-null  float64 
 7   availability_365   38362 non-null  int64   
 8   latitude           38362 non-null  float64 
 9   longitude          38362 non-null  float64 
 10  price_category     38362 non-null  category
dtypes: category(1), float64(5), int64(2), str(3)
memory usage: 5.6 MB


In [36]:
analysis.isnull().sum()

name                 0
neighbourhood        0
room_type            0
price                0
minimum_nights       9
number_of_reviews    0
reviews_per_month    0
availability_365     0
latitude             0
longitude            0
price_category       0
dtype: int64

In [37]:
analysis["price_category"].value_counts()

price_category
$100–199      9862
Under $100    7670
$500+         7090
$300–499      7048
$200–299      6692
Name: count, dtype: int64

In [ ]:
len(neighbourhood_options)

37

In [39]:
neighbourhood_options[:10]

[{'label': 'Alhambra', 'value': 'Alhambra'},
 {'label': 'Beverly Grove', 'value': 'Beverly Grove'},
 {'label': 'Beverly Hills', 'value': 'Beverly Hills'},
 {'label': 'Burbank', 'value': 'Burbank'},
 {'label': 'Culver City', 'value': 'Culver City'},
 {'label': 'Downtown', 'value': 'Downtown'},
 {'label': 'East Hollywood', 'value': 'East Hollywood'},
 {'label': 'East Los Angeles', 'value': 'East Los Angeles'},
 {'label': 'Exposition Park', 'value': 'Exposition Park'},
 {'label': 'Glendale', 'value': 'Glendale'}]

In [41]:
analysis[
    analysis["neighbourhood"].isin(valid_neighbourhoods)
]["neighbourhood"].value_counts().sort_values()

neighbourhood
East Hollywood           310
Marina del Rey           315
Torrance                 316
Topanga                  320
Westlake                 321
Studio City              326
East Los Angeles         351
Malibu                   368
Silver Lake              378
Redondo Beach            379
Lancaster                382
North Hollywood          385
Mid-Wilshire             393
Rowland Heights          416
Manhattan Beach          424
Koreatown                429
Westwood                 432
Beverly Grove            438
Sherman Oaks             448
Woodland Hills           458
Inglewood                464
Mid-City                 532
Burbank                  557
Hollywood Hills          585
Alhambra                 607
Culver City              607
Hollywood Hills West     612
Glendale                 637
Beverly Hills            669
Pasadena                 701
Exposition Park          863
Downtown                 958
Santa Monica            1063
West Hollywood          1147


## 12. Check the interactive groups

Calculate the number of observations available for every control option. Make sure each option returns data and that unexpectedly small groups are understood before building the callback.



In [47]:
# Count listings for every neighborhood offered in the dropdown
group_counts = analysis[
    analysis["neighbourhood"].isin(valid_neighbourhoods)
]["neighbourhood"].value_counts().sort_values()

group_counts

neighbourhood
East Hollywood           310
Marina del Rey           315
Torrance                 316
Topanga                  320
Westlake                 321
Studio City              326
East Los Angeles         351
Malibu                   368
Silver Lake              378
Redondo Beach            379
Lancaster                382
North Hollywood          385
Mid-Wilshire             393
Rowland Heights          416
Manhattan Beach          424
Koreatown                429
Westwood                 432
Beverly Grove            438
Sherman Oaks             448
Woodland Hills           458
Inglewood                464
Mid-City                 532
Burbank                  557
Hollywood Hills          585
Alhambra                 607
Culver City              607
Hollywood Hills West     612
Glendale                 637
Beverly Hills            669
Pasadena                 701
Exposition Park          863
Downtown                 958
Santa Monica            1063
West Hollywood          1147


**What did these checks establish about the data that will appear in the dashboard?**
These checks confirmed that every neighborhood in the dropdown has at least 300 listings in the analysis dataset, so each selection will return data for the charts. However, individual room types within a neighborhood may have fewer listings, so those comparisons should be interpreted carefully.

# Step 5: Evidence



## 13. Create the first static Plotly figure

Build the first figure before placing it in Dash. Give it a descriptive title and human-readable axis and legend labels.



In [48]:
selected_neighbourhood = neighbourhood_options[0]["value"]

filtered = analysis[
    analysis["neighbourhood"] == selected_neighbourhood
]

fig_price = px.box(
    filtered,
    x="room_type",
    y="price",
    color="room_type",
    title=f"Airbnb Prices in {selected_neighbourhood}",
    template="plotly_white",
    height=550,
    labels={
        "room_type": "Room type",
        "price": "Nightly price (USD)"
    }
)

fig_price.update_layout(showlegend=False)

fig_price.show()

**What evidence should a viewer obtain from the first figure?**
The first figure shows how nightly Airbnb prices vary by room type within the selected neighborhood. Viewers can compare median prices, the spread of prices, and unusually expensive listings to help identify which room types may fit their budget.

## 14. Create the second static Plotly figure

The second figure should add different but related evidence rather than repeat the first figure in another style.



In [49]:

fig_availability = px.box(
    filtered,
    x="room_type",
    y="availability_365",
    color="room_type",
    title=f"Airbnb Availability in {selected_neighbourhood}",
    template="plotly_white",
    height=550,
    labels={
        "room_type": "Room type",
        "availability_365": "Days available in the next 365 days"
    }
)

fig_availability.update_layout(showlegend=False)
fig_availability.update_yaxes(range=[0, 365])

fig_availability.show()


**What does the second figure add to the first figure's evidence?**
The second figure shows how availability varies by room type within the selected neighborhood. It adds to the price chart by allowing viewers to compare both nightly prices and the number of days listings are available over the next year. However, it does not show availability for specific travel dates.


## 15. Plan the dashboard



**Complete the dashboard plan before writing the app. Add rows if needed.**

| Component | What it displays or controls | Why it belongs on the dashboard |
|---|---|---|
| Title and introduction | “Airbnb Prices and Availability in Los Angeles,” with a short explanation of the dashboard’s purpose. | Helps visitors understand how the dashboard supports comparing accommodation options. |
| Interactive control | A neighborhood dropdown limited to areas with at least 300 listings. The selection updates both charts. | Lets visitors explore a neighborhood while ensuring each option returns data. |
| First figure | A box plot of nightly prices by room type in the selected neighborhood. | Helps visitors compare typical prices and price variation across room types. |
| Second figure | A box plot of days available over the next 365 days by room type in the selected neighborhood. | Adds availability information to the price comparison. |
| Conclusion and limitation | A brief summary of the patterns shown in the charts and a note that total available days do not establish availability for specific travel dates. | Helps visitors interpret the findings and understand their limits. |



## 16. Create reusable figure code

Write a function that accepts the selected control value, filters or summarizes `analysis`, creates both Plotly figures, and returns the figures in a consistent order. Fix axis ranges or category orders when they need to remain stable across selections.



In [50]:
def create_figures(selected_neighbourhood):

    # Keep listings in the selected neighborhood
    filtered = analysis[
        analysis["neighbourhood"] == selected_neighbourhood
    ]

    # Keep room types in the same order across selections
    room_order = [
        "Entire home/apt",
        "Private room",
        "Shared room",
        "Hotel room"
    ]

    # First chart: nightly prices
    fig_price = px.box(
        filtered,
        x="room_type",
        y="price",
        color="room_type",
        category_orders={"room_type": room_order},
        title=f"Airbnb Prices in {selected_neighbourhood}",
        template="plotly_white",
        height=550,
        labels={
            "room_type": "Room type",
            "price": "Nightly price (USD)"
        }
    )

    fig_price.update_layout(showlegend=False)
    fig_price.update_yaxes(range=[0, analysis["price"].max() * 1.05])

    # Second chart: availability
    fig_availability = px.box(
        filtered,
        x="room_type",
        y="availability_365",
        color="room_type",
        category_orders={"room_type": room_order},
        title=f"Airbnb Availability in {selected_neighbourhood}",
        template="plotly_white",
        height=550,
        labels={
            "room_type": "Room type",
            "availability_365": "Days available in the next 365 days"
        }
    )

    fig_availability.update_layout(showlegend=False)
    fig_availability.update_yaxes(range=[0, 365])

    return fig_price, fig_availability



In [51]:
# previewing function
fig_price, fig_availability = create_figures(
    neighbourhood_options[0]["value"]
)

fig_price.show()
fig_availability.show()

## 17. Build the layout

Create a Dash app and its layout. Include the title, purpose statement, interactive control, two empty `dcc.Graph` components with unique IDs, and short conclusion and limitation statements.



In [56]:
app = Dash(__name__)

app.layout = html.Div(
    [
        html.H1("Airbnb Prices and Availability in Los Angeles"),

        html.P(
            "Compare nightly prices and availability by room type "
            "to help plan your stay in Los Angeles."
        ),

        html.Label("Choose a neighborhood:", htmlFor="neighbourhood-dropdown"),

        dcc.Dropdown(
            id="neighbourhood-dropdown",
            options=neighbourhood_options,
            value=neighbourhood_options[0]["value"],
            clearable=False
        ),

        dcc.Graph(id="price-chart"),

        dcc.Graph(id="availability-chart"),

        html.H3("Conclusion"),
        html.P(
            "In Alhambra, entire homes and apartments generally have "
            "higher nightly prices than private or shared rooms. "
            "Private and shared rooms may offer more affordable options.",
            id="conclusion-text"
        ),

        html.H3("Limitation"),
        html.P(
            "The number of available days does not show whether a listing "
            "is available for your specific travel dates. Although each "
            "neighborhood has at least 300 listings, some room types may "
            "have few listings, making those comparisons less reliable."
        )
    ],
    style={
        "maxWidth": "1000px",
        "margin": "auto",
        "padding": "20px",
        "fontFamily": "Arial"
    }
)


## 18. Connect the callback

Connect the control's `value` to the `figure` property of both graphs. The callback function should call the reusable figure function and return its two figures in the same order as the two outputs.



In [57]:

@app.callback(
    Output("price-chart", "figure"),
    Output("availability-chart", "figure"),
    Input("neighbourhood-dropdown", "value")
)
def update_charts(selected_neighbourhood):
    return create_figures(selected_neighbourhood)


## 19. Check the callback as a Python function

Call the callback or reusable figure function with at least two valid control values. Confirm that each call returns two Plotly figures without an error.



In [58]:
for option in neighbourhood_options[:2]:
    neighbourhood = option["value"]

    fig_price, fig_availability = update_charts(neighbourhood)

    print(f"Neighborhood: {neighbourhood}")
    print(type(fig_price))
    print(type(fig_availability))

    fig_price.show()
    fig_availability.show()


Neighborhood: Alhambra
<class 'plotly.graph_objs._figure.Figure'>
<class 'plotly.graph_objs._figure.Figure'>


Neighborhood: Beverly Grove
<class 'plotly.graph_objs._figure.Figure'>
<class 'plotly.graph_objs._figure.Figure'>


## 20. Run the dashboard

Run the app in a browser tab. Use a port that is not already occupied by another app, and stop the app before restarting the same code.



In [59]:
app.run(jupyter_mode="tab", port=8050)


Dash app running on http://127.0.0.1:8050/


<IPython.core.display.Javascript object>

**After testing every control option, what did you revise or confirm about the dashboard's behavior?**
I confirmed that selecting each neighborhood updated both charts and their titles correctly without errors. The axis ranges and room-type order stayed consistent across selections, making comparisons easier. Every dropdown option returned data, although some neighborhoods had fewer room types represented.


**Which design choice most improves the dashboard's readability or usefulness?**
The neighborhood dropdown most improves the dashboard’s usefulness because it updates both charts together. Visitors can focus on one area and compare prices and availability by room type without being overwhelmed by data from every neighborhood at once.


# Step 6: Conclusion



**State a concise conclusion that directly answers the research question and refers to evidence visible in the dashboard.**
In the Alhambra price chart, entire homes and apartments have a higher median nightly price than private or shared rooms, suggesting that room type matters when choosing affordable accommodation.


**How does the interactive control help the audience examine the evidence?**
The neighborhood dropdown lets visitors select an area and updates both charts together. This helps them compare prices and availability across room types within that neighborhood and explore differences between neighborhoods.


# Step 7: Limitation



**State one important limitation on how the dashboard's evidence should be interpreted.**
The number of available days does not show whether a listing is available for specific travel dates. Visitors should therefore check the listing’s calendar before making booking decisions.

## AI-use reflection



**Describe any AI assistance you used. Identify one suggestion you verified, revised, or rejected. If you did not use AI, state that.**
I used AI to help explain the lab instructions and draft explanations of the dashboard. I rejected its suggestion to use a logarithmic price scale because it was difficult for me to explain. I kept the regular price scale and used simpler formatting to improve readability.


## Submission checklist

Before submitting, confirm that:

- [ ] every prompt has one complete response;
- [ ] all notebook cells run in order without an error;
- [ ] the dashboard opens and every control option works;
- [ ] the two figures use readable titles, labels, categories, and colors;
- [ ] comparable views retain stable axes or category orders when needed;
- [ ] the dataset source is cited and the required data file accompanies the submission;
- [ ] the notebook and dashboard code are organized and readable;
- [ ] the conclusion and limitation are supported by the displayed evidence; and
- [ ] the work is submitted by the deadline.

